pkg/apis/policy/v1alpha1/upstreamtrafficsetting.go

```diff
@@ -99,6 +113,111 @@ type HTTPConnectionSettings struct {
 }
 
+// RateLimitSpec defines the rate limiting specification for
+// the upstream host.
+type RateLimitSpec struct {
+	// Local specified the local rate limiting specification
+	// for the upstream host.
+	// Local rate limiting is enforced directly by the upstream
+	// host without any involvement of a global rate limiting service.
+	// This is applied as a token bucket rate limiter.
+	// +optional
+	Local *LocalRateLimitSpec `json:"local,omitempty"`
+}
+
+// LocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host.
+type LocalRateLimitSpec struct {
+	// TCP defines the local rate limiting specification at the network
+	// level. This is a token bucket rate limiter where each connection
+	// consumes a single token. If the token is available, the connection
+	// will be allowed. If no tokens are available, the connection will be
+	// immediately closed.
+	// +optional
+	TCP *TCPLocalRateLimitSpec `json:"tcp,omitempty"`
+
+	// HTTP defines the local rate limiting specification for HTTP traffic.
+	// This is a token bucket rate limiter where each request consumes
+	// a single token. If the token is available, the request will be
+	// allowed. If no tokens are available, the request will receive the
+	// configured rate limit status.
+	HTTP *HTTPLocalRateLimitSpec `json:"http,omitempty"`
+}
+
+// TCPLocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host at the TCP level.
+type TCPLocalRateLimitSpec struct {
+	// Connections defines the number of connections allowed
+	// per unit of time before rate limiting occurs.
+	Connections uint32 `json:"connections"`
+
+	// Unit defines the period of time within which connections
+	// over the limit will be rate limited.
+	// Valid values are "second", "minute" and "hour".
+	Unit string `json:"unit"`
+
+	// Burst defines the number of connections above the baseline
+	// rate that are allowed in a short period of time.
+	// +optional
+	Burst uint32 `json:"burst,omitempty"`
+}
+
+// HTTPLocalRateLimitSpec defines the local rate limiting specification
+// for the upstream host at the HTTP level.
+type HTTPLocalRateLimitSpec struct {
+	// Requests defines the number of requests allowed
+	// per unit of time before rate limiting occurs.
+	Requests uint32 `json:"requests"`
+
+	// Unit defines the period of time within which requests
+	// over the limit will be rate limited.
+	// Valid values are "second", "minute" and "hour".
+	Unit string `json:"unit"`
+
+	// Burst defines the number of requests above the baseline
+	// rate that are allowed in a short period of time.
+	// +optional
+	Burst uint32 `json:"burst,omitempty"`
+
+	// ResponseStatusCode defines the HTTP status code to use for responses
+	// to rate limited requests. Code must be in the 400-599 (inclusive)
+	// error range. If not specified, a default of 429 (Too Many Requests) is used.
+	// +optional
+	ResponseStatusCode uint32 `json:"responseStatusCode,omitempty"`
```

Comment:
Does it make sense for this to be a pointer so that presence is easily validated?

Reply:
I thought about this, and since Envoy will default to 429 anyway for values < 400, keeping this an uint32 should be ok.

Reply:
I think we should opt more for default values in the CRD definition. That way when a user performs a Get on an object, they see the value.

This is what K8s recommends for default values, and ie: what they use to populate objects without a specified namespace

Reply:
In some cases, it helps to let Envoy use its default when fields are unset, and we can simply document these instead of encoding Envoy defaults throughout the code. I understand your point though, where setting defaults in the CRD may be more intuitive.

Reply:
that makes sense to!

Reply:
Yeah I can see both sides. I think if we're going to allow Envoy defaults, we just need to document them and specify that Envoy will handle it.